### ISSUE TITLE
Baseball Simulator 1.000: broken after pause and field change

### ISSUE THREAD

**Issue by Kopert** [MiSTer-devel/NES_MiSTer#278]
**Core**: NES_20211018.rbf
**ROM**: Baseball Simulator 1.000 (USA).nes, No-Intro, MD5: 4614fd4608ef5f04f05330f4fc65d239

**Reproduction steps**:
1. Start game
2. Press start, skip messages about backup being lost (SRAM issue?)
3. Choose Exhibition, 1Player, default options for first and second team, default choice for stadium
4. Push start
5. Pick default choice for pitcher, choose yes to pick him, then choose "use" to select the lineup and start the game
6. After "play ball", as soon as the game starts and before the pitcher pitches, press Start to pause ("time" shows up on screen)
7. Note that, on the left, it shows "BO JOSE 0" and, on the right, it shows "BO 1GLEN .367 Hr 6", as normal
8. Press A to enter the pitch hitter screen
9. Navigate to "Cancel" and press A
10. **Now it is broken**! On the right it now shows "00 0GLEN .000 Hr 0" for the pitcher (which is on the opposite team and shouldn't have changed, or have completely broken stats) and he WON'T PITCH. He also changes his uniform color and is sweating (tired) and the batter switches to the left. The game apparently is now 2 players; if you swap the joysticks, you can pitch for the opposite team (since the pitcher no longer does it on his own).

Expected behavior: the game continues as before the pause.

Running the game on an emulator the messages about the backup being lost do not show up and, upon running the exacty same reproduction steps above, after step 9 the game resumes as normal from where it stopped.

My guess: there's a problem with SRAM on this game which is preventing proper access to roster data, causing these issues. Maybe. Just a guess, tho, I don't really know. I tried manually saving the game from the OSD, which generated an 8kb file on the /media/fat/saves/NES folder, but none of the behavior changed. The User led never lights up while running the game.

**Comment by SegaSnatcher** [comment]
If you want you can make it easier and provide a save state too.

**Comment by Kopert** [comment]
Oh, sure, anything to help. [Here's a savestate](https://github.com/MiSTer-devel/NES_MiSTer/files/7535688/Baseball.Simulator.1.000.USA._1.ss.zip). If you load this, it will be paused just before step 8 on my reproduction steps above.

From there, press **A**, then **Up**, then **A** (steps 8 and 9) to cause the error.

**Comment by SegaSnatcher** [comment]
I tried to reproduce this issue, but the game is working fine for me.  I also don't get a message about backup being lost. 

My rom information from 2.0 rom pack.  

MD5: 8CFE109D9CA1CE97A5775876F54D5E5F

**Comment by SegaSnatcher** [comment]
OK, so I tried your save state and was able to reproduce your issue, but if I do everything manually I don't get this issue, interesting.

**Comment by Kopert** [comment]
Those are indeed different roms. I tested with 8cfe109d9ca1ce97a5775876f54d5e5f and it worked fine on my end.

I tested the Virtual Console rom from the same No-Intro pack I used for my main rom (940d6a25c9df1df7c391c8ef1fb82d8f) and the issue occurs there as well. All roms work fine on emulators.

I can offer links to archive.org to the problematic roms but I don't know if that would be ok. If you search "archive.org no-intro nes" it should be the top result.

**Comment by SegaSnatcher** [comment]
One thing I noticed about your save state.  It already changes the color of the batter and shows BOS as 00 before I go into the Pitcher menu.

**Comment by Kopert** [comment]
> 
> 
> One thing I noticed about your save state. It already changes the color of the batter and shows BOS as 00 before I go into the Pitcher menu. Try your save state on the good working rom.

Uh, not for me. I loaded the bad savestate on the good rom and it looks the same, correct info etc. And it doesn't break after going into the menu.

I don't know if this is useful, I loaded both roms into a mapper reader and the only differences I could spot were headers.

Problematic rom: `NES.......w.....` `4E 45 53 1A 08 10 12 08 00 00 77 07 00 00 00 00 `
Good rom: `NES.......p.....` `4E 45 53 1A 08 10 12 08 00 00 70 00 00 00 00 01 `

**Comment by Kopert** [comment]
I found a version of the rom (MD5 `cdf5a1040fe782d6a308c495d6b6a0b1`) without NES 2.0 headers and it also worked fine.

**Comment by Kopert** [comment]
I'm trying to figure out what the difference is in the NES 2.0 headers that makes `4614fd4608ef5f04f05330f4fc65d239 ` not work on the MiSTer but lets it work fine on emulators. Most No-Intro packs I managed to locate have this version of the rom, and it worries me that other games might have problems because of these headers.

**Comment by Kopert** [comment]
I'm comparing the working rom with the non-working one with a hex editor. The ONLY differences are in the NES 2.0 headers (a rom with basic iNES headers works fine; all these roms work fine in an emulator).

The only differences are in offsets 10, 11 and 15. I'm trying to follow `[this documentation](https://wiki.nesdev.org/w/index.php/NES_2.0)` to figure things out. Below are my best guesses based on my interpretation of the documentation:

Offset 10 is PRG-RAM/EEPROM size.
Correct rom is `0x70`; that means there is no PRG-RAM and PRG-NVRAM/EEPROM is 8192 bytes.
Wrong rom is `0x77`; that means PRG-RAM and PRG-NVRAM/EEPROM are both 8192 bytes.

Offset 11 is CHR-RAM size.
Correct rom is `0x00`; that means there is no CHR-(NV)RAM.
Wrong rom is `0x07`; that means that CHR-RAM is 8192 bytes and there is no CHR-NVRAM.

Offset 15 is Default Expansion Device; I don't know how to interpret this one.
Correct rom is `0x01`
Wrong rom is `0x00`

I hope this helps track down the issue.

**Comment by paulb-nl** [comment]
The NES core relies on correct header data so weird things can happen if the header is wrong.

Software emulators usually ignore the header and use their own database. Try to load a rom where the header is all zeros or keep the first 4 bytes only.

**Comment by Kopert** [comment]
> 
> 
> The NES core relies on correct header data so weird things can happen if the header is wrong.
> 
> Software emulators usually ignore the header and use their own database. Try to load a rom where the header is all zeros or keep the first 4 bytes only.

I got an unheadered version of the same rom and tried loading it up on an emulator and it seems you're right, it just worked.

I've come to conclude that the current approach is indeed correct and I retract my request to make the core "more compatible". I agree that it is better and more accurate to depend on correct headers. The core already has a reasonable fallback for iNES headers and that should be enough.

Thanks to everyone that commented on this issue. Based on everything I've learned I'm closing it.
